torchdata/stateful_dataloader/stateful_dataloader.py

```diff
@@ -861,18 +870,30 @@ def __init__(self, loader, next_iter_state):
         _utils.signal_handling._set_worker_pids(id(self), tuple(w.pid for w in self._workers))  # type: ignore[misc]
         _utils.signal_handling._set_SIGCHLD_handler()
         self._worker_pids_set = True
-        self._snapshot, self._worker_snapshots, self._main_snapshots = {}, {}, collections.deque()  # type: ignore[var-annotated]
-
-        self._main_state_0 = self._get_main_state()
+        self._snapshot, self._main_snapshots = {}, collections.deque()  # type: ignore[var-annotated]
+        # NOTE [ Incremental Worker State ]
+        # We only send deltas between incremental worker state to the main process. We synchronize
+        # the initial states on worker startup, when it sends an _AckStartup signal back with the initial
+        # worker states, and if persistent_workers is True, then the worker sends back an initial
+        # state after acking the _ResumeIteration signal.
+        #
+        # We need to send initial worker state back to the main process to handle state_dict() requests
+        # before n >= num_workers steps are taken.
+        self._worker_snapshots: Dict[str, _IncrementalWorkerState] = {}
         self._reset(loader, first_iter=True, prime_prefetch=next_iter_state is None)
 
         # Try to restore main state
         if next_iter_state is not None:
             self._restore_main_state(next_iter_state[self._SNAPSHOT][self._MAIN_SNAPSHOT])
             self._num_yielded = next_iter_state[self._SNAPSHOT][self._SNAPSHOT_STEP]
 
-            # Back-fill the worker snapshots before starting, in case of failure before a full cycle
-            self._worker_snapshots = {key: _IncrementalWorkerState(state) for key, state in worker_states.items()}
+            self._update_snapshot(
+                snapshot_step=next_iter_state[self._SNAPSHOT][self._SNAPSHOT_STEP],
+                last_yielded_worker_id=next_iter_state[self._SNAPSHOT][self._LAST_YIELDED_WORKER_ID],
+                num_workers=self._num_workers,
+                main_snapshot=next_iter_state[self._SNAPSHOT][self._MAIN_SNAPSHOT],
+                worker_snapshots=self._worker_snapshots,
```

Comment:
where is the worker states extracted from the next_iter_state?

Reply:
They're initialized in `_reset`